UnityProject/Assets/Scripts/Weapons/Projectiles/Behaviours/ProjectileAnimatedDespawn.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System.Collections;
+using UnityEngine;
+
+namespace Weapons.Projectiles.Behaviours
+{
+	/// <summary>
+	/// Creates an animated tile at a point of collision or end
+	/// After animation, applies old tile effect and despawns gameobject
+	/// </summary>
+	public class ProjectileAnimatedDespawn : MonoBehaviour, IDespawn
+	{
+		[SerializeField] private AnimatedTile animatedTile;
+
+		[Tooltip("Living time of animated tile.")]
+		[SerializeField] private float animationTime;
+
+		public void Despawn(RaycastHit2D hit, Vector2 point)
+		{
+			if (hit.collider == null)
+			{
+				OnBeamEnd(point);
+			}
+			else
+			{
+				OnCollision(hit);
+			}
+		}
+
+		private void OnBeamEnd(Vector2 distance)
+		{
+			var layerMetaTile = GetComponentInParent<MetaTileMap>();
+			var cellPos = layerMetaTile.WorldToCell(distance);
+			var tileChangeManager = GetComponentInParent<TileChangeManager>();
+
+			ProcessAnimation(layerMetaTile, cellPos, tileChangeManager);
+		}
+
+		private void OnCollision(RaycastHit2D hit)
+		{
+			var coll = hit.collider;
+			var layerMetaTile = coll.GetComponentInParent<MetaTileMap>();
+
+			var bulletHitTarget = Vector3.zero;
+			bulletHitTarget.x = hit.point.x - 0.01f * hit.normal.x;
+			bulletHitTarget.y = hit.point.y - 0.01f * hit.normal.y;
+			Vector3Int cellPos = layerMetaTile.WorldToCell(bulletHitTarget);
+
+			var tileChangeManager = coll.GetComponentInParent<TileChangeManager>();
+
+			ProcessAnimation(layerMetaTile, cellPos, tileChangeManager);
+		}
+
+		private void ProcessAnimation(MetaTileMap layerMetaTile, Vector3Int cellPos, TileChangeManager tileChangeManager)
+		{
+			var oldEffectLayerTile = layerMetaTile.GetTile(cellPos, LayerType.Effects);
+
+			StartCoroutine(EnableAnimationCoroutine(cellPos, tileChangeManager, oldEffectLayerTile));
+		}
+
+		private IEnumerator EnableAnimationCoroutine(Vector3Int cellPos, TileChangeManager tileChangeManager, LayerTile oldEffectLayerTile)
+		{
+			tileChangeManager.UpdateTile(cellPos, TileType.Effects, animatedTile.name);
```

Comment:
did you check if modifying cellPos z would mitigate the need to remove and reapply old effect?
I know matrices are 3d but don't remember if out network tile sync supports custom z values

Reply:
I didn't check. Didn't dig this logic further tbh, just fixed the issues Proto-Kinetic-Accelerator had with this animation.